### PR TITLE
Golden hearts probabilities

### PR DESCRIPTION
**Pull request by CameroonD**
Golden hearts act like shields for your base health, remaining even when possessing an enemy.
They also restore your base health to maximum.
Golden hearts are the rarest of hearts with half gold hearts having a probability (if anything is dropped at all) of 15% and full gold hearts of 10%.

### CODE REVIEW

File: Assets/Scripts/Entity/Enemy/EnemyDrop.cs [hunk 1]
```diff
@@ -26,12 +35,34 @@ void SubscribeToEnemyDeath()
         }
     }
 
+    void PopulateDropList()
+    {
+        for(int i = 0; i < heartPrefabs.Count; i++)
```
Comment (Jazy5552) [inline Assets/Scripts/Entity/Enemy/EnemyDrop.cs:40]: Hello sir, have you heard of our lord and savior, the [Foreach loop](https://www.geeksforgeeks.org/c-sharp-foreach-loop/)? :)

File: Assets/Scripts/Entity/Enemy/EnemyDrop.cs [hunk 2]
```diff
@@ -26,12 +35,34 @@ void SubscribeToEnemyDeath()
...
+        {
+            var dropObject = new Drop
```
Comment (Jazy5552) [inline Assets/Scripts/Entity/Enemy/EnemyDrop.cs:42]: Excuse me? `var`? What is this, Visual Basic (or god forbid JS)? Use static types like a real man.

File: Assets/Scripts/Entity/Enemy/Heart.cs [hunk 3]
```diff
@@ -5,21 +5,40 @@
 public class Heart : MonoBehaviour
 {
     public int Health;
+    public bool Golden;
 
     private void OnCollisionEnter(Collision other)
     {
         if (other.gameObject.CompareTag("Player"))
```
Comment (Jazy5552) [inline Assets/Scripts/Entity/Enemy/Heart.cs:12]: I see you haven't heard of the glorious [Fail-Fast](https://en.wikipedia.org/wiki/Fail-fast).

tl;dr: `if (!other.gameObject.CompareTag("Player")) return;`

**Review by Jazy5552 — APPROVED** [review]
Looks _alright_

File: Assets/Scripts/Entity/Enemy/EnemyDrop.cs (hunk 1 above)
Comment (CameroonD) [reply Assets/Scripts/Entity/Enemy/EnemyDrop.cs:40]: Yes but I needed the index so I could access elements of both lists at the same time.

File: Assets/Scripts/Entity/Enemy/Heart.cs (hunk 3 above)
Comment (CameroonD) [reply Assets/Scripts/Entity/Enemy/Heart.cs:12]: Boi now you criticising lines of code I didn't even make in this PR lol.

File: Assets/Scripts/Entity/Enemy/EnemyDrop.cs (hunk 2 above)
Comment (CameroonD) [reply Assets/Scripts/Entity/Enemy/EnemyDrop.cs:42]: var master race. JS will rule us all. I am of the opinion that in C# you only ever need to explicitly declare type if it could possibly be ambiguous. In this case you can see on the same line exactly what the type will be. Additionally var isn't like var in JS, it's not an actual type and it is converted into the corresponding type at compile time, not runtime which means there is no performance hit.

Comment (Jazy5552) [reply Assets/Scripts/Entity/Enemy/EnemyDrop.cs:42]: Biggest (and I think only) issue with `var` is readability. I absolutely detest JS because when you get large chunks of code, unless the variable names are really verbose, you have no idea what each variable is. In fact, you pretty accept the fact that you don't know what a variable is and just "assume" it's of a type you can work with. This is what makes JS so bad. People come in and just start doing what ever they want, adding properties to objects at runtime, adding methods to objects at runtime, that by the time you get the object, only god knows what is in it and where it has been.

JS aside, `var` is still bad because of the reason for large chunks of code. Sure here, it can be obvious that `dropObject = new Drop` is a Drop object. But there simply is no need, you literally save typing 1 letter, and you lose a good deal of readability. Specially if you work with more complex design patterns such as factories, or you use polymorphism. 

General rule of thumb, you avoid `var`/`auto` if possible. You never gain anything, and you will usually lose quite a bit of readability.

File: Assets/Scripts/Entity/Enemy/Heart.cs (hunk 3 above)
Comment (AleksandrHovhannisyan) [reply Assets/Scripts/Entity/Enemy/Heart.cs:12]: lol
